File: src/main/java/com/hileco/cortex/instructions/ProgramRunner.java
package com.hileco.cortex.instructions;

import com.hileco.cortex.vm.VirtualMachine;
import com.hileco.cortex.instructions.ProgramException.Reason;

public class ProgramRunner {

    private final VirtualMachine virtualMachine;

    public ProgramRunner(VirtualMachine virtualMachine) {
        this.virtualMachine = virtualMachine;
    }

    @SuppressWarnings("unchecked")
    public void run() throws ProgramException {
        var programContext = this.virtualMachine.getPrograms().peek();
        while (programContext.getInstructionPosition() != programContext.getProgram().getInstructions().size()) {
            var currentInstructionPosition = programContext.getInstructionPosition();
            var current = programContext.getProgram().getInstructions().get(currentInstructionPosition);
            current.execute(this.virtualMachine, programContext);
            programContext = this.virtualMachine.getPrograms().peek();
            if (programContext == null) {
                break;
            }
            if (programContext.getInstructionPosition() == currentInstructionPosition) {
                programContext.setInstructionPosition(currentInstructionPosition + 1);
            }
            programContext.setInstructionsExecuted(programContext.getInstructionsExecuted() + 1);
            if (programContext.getInstructionsExecuted() >= programContext.getInstructionLimit()) {
                throw new ProgramException(programContext, Reason.INSTRUCTION_LIMIT_REACHED_ON_PROGRAM_LEVEL);
            }
            this.virtualMachine.setInstructionsExecuted(this.virtualMachine.getInstructionsExecuted() + 1);
            if (this.virtualMachine.getInstructionsExecuted() >= this.virtualMachine.getInstructionLimit()) {
                throw new ProgramException(programContext, Reason.INSTRUCTION_LIMIT_REACHED_ON_PROCESS_LEVEL);
            }
        }
    }
}

File: src/main/java/com/hileco/cortex/instructions/ProgramRunner.kt
package com.hileco.cortex.instructions

import com.hileco.cortex.instructions.ProgramException.Reason
import com.hileco.cortex.vm.ProgramConstants.Companion.INSTRUCTION_LIMIT
import com.hileco.cortex.vm.concrete.ProgramContext
import com.hileco.cortex.vm.concrete.VirtualMachine

class ProgramRunner(private val virtualMachine: VirtualMachine) {
    fun run() {
        if (virtualMachine.programs.isEmpty()) {
            return
        }
        var context: ProgramContext = virtualMachine.programs.peek()
        while (context.instructionPosition < context.program.instructions.size) {
            val currentInstructionPosition = context.instructionPosition
            val current = context.program.instructions[currentInstructionPosition]
            current.execute(virtualMachine, context)
            if (virtualMachine.programs.isEmpty()) {
                break
            }
            context = virtualMachine.programs.peek()
            if (context.instructionPosition == currentInstructionPosition) {
                context.instructionPosition = currentInstructionPosition + 1
            }
            context.instructionsExecuted++
            if (context.instructionsExecuted >= INSTRUCTION_LIMIT) {
                throw ProgramException(Reason.INSTRUCTION_LIMIT_REACHED_ON_PROGRAM)
            }
            virtualMachine.instructionsExecuted++
            if (virtualMachine.instructionsExecuted >= INSTRUCTION_LIMIT) {
                throw ProgramException(Reason.INSTRUCTION_LIMIT_REACHED_ON_VIRTUAL_MACHINE)
            }
        }
    }
}
